contracts/EpochTokenLocker.sol

```diff
@@ -0,0 +1,146 @@
+pragma solidity ^0.5.0;
+
+import "openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+import "openzeppelin-solidity/contracts/math/Math.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+
+// EpochTokenLocker saveguards tokens for applications with constant-balances during discrete epochs
+// It allows to deposit token which become credited in the next epoch and allows to request a token-withdraw
+// which becomes claimable after the current epoch expired.
+
+contract EpochTokenLocker {
+    using SafeMath for uint;
+
+    event Deposit(
+        address user,
+        address token,
+        uint amount,
+        uint stateIndex
+    );
+
+    event withdrawRequest(
+        address user,
+        address token,
+        uint amount,
+        uint stateIndex
+    );
+
+    event Withdraw(
+        address user,
+        address token,
+        uint amount
+    );
+
+    // User => Token => BalanceState
+    mapping(address => mapping(address => BalanceState)) private balanceStates;
+
+    struct BalanceState {
+        uint256 balance;
+        PendingFlux pendingDeposits; // deposits will be credited in any next epoch, i.e. currentStateIndex > stateIndex
+        PendingFlux pendingWithdraws; // withdraws are allowed in any next epoch , i.e. currentStateIndex > stateIndex
+    }
+
+    struct PendingFlux {
+        uint256 amount;
+        uint256 stateIndex;
+    }
+
+    uint256 public currentStateIndex = 0;
+
+    function updateAndGetBalance(address user, address token) public returns(uint256) {
+        updateDepositsBalance(user, token);
+        uint balance = balanceStates[user][token].balance;
+        if (balanceStates[user][token].pendingWithdraws.stateIndex < currentStateIndex) {
+            balance -= Math.min(balanceStates[user][token].pendingWithdraws.amount, balance);
+        }
+        return balance;
+    }
+
+    function deposit(address token, uint amount) public {
+        updateDepositsBalance(msg.sender, token);
```

Comment:
Who decides which tokens can be deposited? Do we want to limit this here or only on the trading level?

Reply:
I think it's good that the concerns are separated. It can be a specialization of this contract. One reason is because the list of tokens could have different strategies/implementation on how this whitelisted tokens are maintained